### PR TITLE
Add syntax_suggest as default gem

### PR DESCRIPTION
**Pull request by schneems**
Adds the `syntax_suggest` syntax error display tool to Ruby through the same mechanism as `error_highlight` and `did_you_mean`. Reference ticket: https://bugs.ruby-lang.org/issues/18159

close #4845

## What is syntax_suggest?

When a syntax error is raised by requiring a file, dead_end will use a combination of indentation and lexing to identify the problem.

> Note: Previously this tool was named `dead_end`. 

## Known issues

- SyntaxSearch's approach of showing syntax errors only works through integration with `require`, `load`, `autoload`, and `require_relative` (since it monkeypatches them to detect syntax errors). It does not work with direct Ruby file invocations https://github.com/zombocom/dead_end/issues/31.
  - This causes failure in the test suite (test_expected_backtrace_location_when_inheriting_from_basic_object_and_including_kernel) and confusion when inspecting backtraces if there's a different error when trying to require a file such as measuring memory (https://github.com/zombocom/syntax_suggest/issues/124#issuecomment-1006705016).
  - Discussed fix. We previously talked about opening up `SyntaxError` to be monkeypatched in the same way that other gems hook into `NoMethodError`. This is currently not possible and requires development work. When we last talked about it at RubyKaigi Nobu expressed an ability to make such a change.

### CODE REVIEW

**Comment by schneems** [comment]
This is ready for discussion. To fix the failing tests we would have to make SyntaxError monkeypatchable and then update DeadEnd to use the new interface. They are currently:

```
  Retrying...
  
    1) Failure:
  TestGemRequire#test_expected_backtrace_location_when_inheriting_from_basic_object_and_including_kernel [/home/runner/work/ruby/ruby/src/test/rubygems/test_require.rb:667]:
  Expected /main\.rb:2: warning: This is a deprecated method$/ to match "/home/runner/work/ruby/ruby/src/lib/dead_end/core_ext.rb:12: warning: method redefined; discarding old load\n" +
  "/home/runner/work/ruby/ruby/src/lib/dead_end/core_ext.rb:18: warning: method redefined; discarding old require\n" +
  "/home/runner/work/ruby/ruby/src/lib/dead_end/core_ext.rb:24: warning: method redefined; discarding old require_relative\n" +
  "/home/runner/work/ruby/ruby/src/lib/dead_end/core_ext.rb:19: warning: This is a deprecated method\n".
  
    2) Failure:
  TestGemRequire#test_no_kernel_require_in_warn_with_uplevel [/home/runner/work/ruby/ruby/src/test/rubygems/test_require.rb:613]:
  Expected /main\.rb:1: warning: uplevel\ntest\n$/ to match "/home/runner/work/ruby/ruby/src/lib/dead_end/core_ext.rb:12: warning: method redefined; discarding old load\n" +
  "/home/runner/work/ruby/ruby/src/lib/dead_end/core_ext.rb:18: warning: method redefined; discarding old require\n" +
  "/home/runner/work/ruby/ruby/src/lib/dead_end/core_ext.rb:24: warning: method redefined; discarding old require_relative\n" +
  "/home/runner/work/ruby/ruby/src/lib/dead_end/core_ext.rb:19: warning: uplevel\n" +
  "test\n".
  
    3) Failure:
  TestGemRequire#test_no_kernel_require_in_Kernel_warn_with_uplevel [/home/runner/work/ruby/ruby/src/test/rubygems/test_require.rb:613]:
  Expected /main\.rb:1: warning: uplevel\ntest\n$/ to match "/home/runner/work/ruby/ruby/src/lib/dead_end/core_ext.rb:12: warning: method redefined; discarding old load\n" +
  "/home/runner/work/ruby/ruby/src/lib/dead_end/core_ext.rb:18: warning: method redefined; discarding old require\n" +
  "/home/runner/work/ruby/ruby/src/lib/dead_end/core_ext.rb:24: warning: method redefined; discarding old require_relative\n" +
  "/home/runner/work/ruby/ruby/src/lib/dead_end/core_ext.rb:19: warning: uplevel\n" +
  "test\n".
  
  Finished tests in 240.244681s, 90.4703 tests/s, 11528.7964 assertions/s.
  21735 tests, 2769732 assertions, 3 failures, 0 errors, 100 skips
  
  ruby -v: ruby 3.2.0dev (2022-04-28T18:55:26Z pull/5859/merge 2f7eed2985) [x86_64-linux]
  make: *** [uncommon.mk:823: yes-test-all] Error 3
  Error: Process completed with exit code 2.
```

File: gem_prelude.rb [hunk 1]
```diff
@@ -17,3 +17,9 @@
 rescue LoadError
   warn "`did_you_mean' was not loaded."
 end if defined?(DidYouMean)
+
+begin
+  require 'dead_end'
```
Comment (nobu) [inline gem_prelude.rb:22]:
```suggestion
  require 'dead_end/core_ext'
```

File: lib/dead_end/core_ext.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
+# frozen_string_literal: true
+
```
Comment (nobu) [inline lib/dead_end/core_ext.rb:2]:
```suggestion
require_relative "version"

module DeadEnd
  class << self
    def handle_error(e)
      require_relative "api"
      handle_error(e)
    end

    alias handle_error handle_error
  end
end

if ::SyntaxError.method_defined?(:detailed_message)
  module Handler
    def detailed_message(highlight: nil, **)
      if highlight and /\A(?:require(?:_relative)?|load)\z/.match?(backtrace_locations.first.label)
        DeadEnd.handle_error(self)
      end
    rescue SystemStackError => e
      STDERR.puts e
    ensure
      super
    end

    ::SyntaxError.prepend(self)
  end

  return
end

```

**Review by nobu — COMMENTED** [review]
DeadEnd is expected not to run when no syntax error.
It would be better to load lazily.

File: gem_prelude.rb (hunk 1 above)
Comment (schneems) [reply gem_prelude.rb:22]: I'm baking this ability directly into the library with https://github.com/zombocom/dead_end/pull/142

File: lib/dead_end/core_ext.rb (hunk 2 above)
Comment (schneems) [reply lib/dead_end/core_ext.rb:2]: Updated in https://github.com/zombocom/dead_end/pull/139.

**Comment by schneems** [comment]
All tests are now passing. I believe this is ready for review.

File: lib/dead_end/core_ext.rb [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+# Allow lazy loading, only load code if/when there's a syntax error
+autoload :DeadEnd, "dead_end/api"
+
+# Ruby 3.2+ has a cleaner way to hook into Ruby that doesn't use `require`
+if SyntaxError.new.respond_to?(:detailed_message)
```
Comment (nobu) [inline lib/dead_end/core_ext.rb:7]: A new instance will not be needed.
```suggestion
if SyntaxError.method_defined?(:detailed_message)
```

File: lib/dead_end/api.rb [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+  end
+
+  # DeadEnd.call [Private]
+  #
+  # Main private interface
+  def self.call(source:, filename: DEFAULT_VALUE, terminal: DEFAULT_VALUE, record_dir: DEFAULT_VALUE, timeout: TIMEOUT_DEFAULT, io: $stderr)
+    search = nil
+    filename = nil if filename == DEFAULT_VALUE
+    Timeout.timeout(timeout) do
+      record_dir ||= ENV["DEBUG"] ? "tmp" : nil
```
Comment (nobu) [inline lib/dead_end/api.rb:70]: "DEBUG" feels too generic to me.

File: lib/dead_end/api.rb [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+  # Used to generate a unique directory to record
+  # search steps for debugging
+  def self.record_dir(dir)
+    time = Time.now.strftime("%Y-%m-%d-%H-%M-%s-%N")
+    dir = Pathname(dir)
+    symlink = dir.join("last").tap { |path| path.delete if path.exist? }
+    dir.join(time).tap { |path|
+      path.mkpath
+      FileUtils.symlink(path.basename, symlink)
+    }
```
Comment (nobu) [inline lib/dead_end/api.rb:98]: Why don't use `ln_sf`?

```suggestion
    dir.join(time).mkpath
    FileUtils.ln_sf(time, dir.join("last"))
```

File: lib/dead_end/core_ext.rb [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+  module DeadEndUnloaded
```
Comment (nobu) [inline lib/dead_end/core_ext.rb:8]: Why isn't this `UnloadedDeadEnd` in dead_end/version.rb?

File: lib/dead_end/core_ext.rb [hunk 7]
```diff
@@ -0,0 +1,88 @@
+# frozen_string_literal: true
+
+# Allow lazy loading, only load code if/when there's a syntax error
+autoload :DeadEnd, "dead_end/api"
```
Comment (nobu) [inline lib/dead_end/core_ext.rb:4]: I doubt if this works after `DeadEnd` is defined in `ruby_opt_init`.
`autoload` does nothing for already defined constant.

File: lib/dead_end/core_ext.rb (hunk 3 above)
Comment (schneems) [reply lib/dead_end/core_ext.rb:7]: Thank you for your PR. I will pull that change in on the next commit.

File: lib/dead_end/api.rb (hunk 5 above)
Comment (schneems) [reply lib/dead_end/api.rb:98]: Thank you for your PR. I will pull that change in on the next commit.

File: lib/dead_end/core_ext.rb (hunk 7 above)
Comment (schneems) [reply lib/dead_end/core_ext.rb:4]: Good to know. I will try to switch to a different constant.

File: lib/dead_end/core_ext.rb (hunk 6 above)
Comment (schneems) [reply lib/dead_end/core_ext.rb:8]: When developing the lazy load feature this constant came first, the constant in the version.rb was only needed when the gemspec gets loaded and was not discovered until later. I never went back and updated this file to depend on the declaration in version.rb but I can make that update.

File: lib/dead_end/core_ext.rb (hunk 7 above)
Comment (schneems) [reply lib/dead_end/core_ext.rb:4]: Here is my new approach/proposal https://github.com/zombocom/dead_end/pull/148

File: lib/dead_end/api.rb (hunk 4 above)
Comment (schneems) [reply lib/dead_end/api.rb:70]: It is mostly for internal/library debugging anyway. The expected use case of outputting intermediate files would be to use the `dead_end` cli along with the `--record` flag.

This was never a documented interface so it is not a breaking change. Something like `DEAD_END_DEBUG` is okay. I will make the update.

**Comment by schneems** [comment]
I've rebased to change the name of the project from `DeadEnd` to `SyntaxSuggest`. All checks are passing except for "Code scanning results / CodeQL" which doesn't appear to be related to my code change. I've got no other pending changes at this time but in https://bugs.ruby-lang.org/issues/18159 we talked about a new API for providing source code strings and location. If those are implemented then I can update my codebase to use them.

**Pull request merged**